### ISSUE TITLE
kernel threadgroup size limit

### ISSUE THREAD

**Issue by funmania** [ml-explore/mlx-swift-examples#104]
When sorting entries on argSort(array, axis: -1, stream: .gpu), sys crashes 

-[MTLDebugComputeCommandEncoder _validateThreadsPerThreadgroup:]:1267: failed assertion `(threadsPerThreadgroup.width(929) * threadsPerThreadgroup.height(1) * threadsPerThreadgroup.depth(1))(929) must be <= 256. (kernel threadgroup size limit)'

**Comment by awni** [comment]
Which example are you running / what is the shape of the input to argSort?

**Comment by davidkoski** [comment]
Also, what system?  iOS?  Simulator?

**Comment by funmania** [comment]
I was running a example code 
` let randArr = MLXRandom.randInt(0 ..< 100, [1900000])
   let sortIndexes = argSort(randArr, axis: -1, stream: .gpu)`
and was getting the crash, suprisingly its not consistent, I have tried it few times, initially I was getting crash but now its working.

I was running it on M1 Max 32 gb, Mac simulator (Designed for iPad)

**Comment by awni** [comment]
Interesting.. I can't reproduce it in MLX core with Metal validation enabled. Is it crashing intermittently now or not at all?

**Comment by funmania** [comment]
Its crashing when I try to print the sorted array, otherwise not.
`let randArr = MLXRandom.randInt(0 ..< 100, [1900000])`
`let sortIndexes = argSort(randArr, axis: -1, stream: .gpu)`
`let sorted = randArr[sortIndexes]`
`print(sorted)`

**Comment by funmania** [comment]
![Uploading Screenshot 2024-08-11 at 9.46.48 PM.png…]()

**Comment by funmania** [comment]
<img width="1727" alt="Screenshot" src="https://github.com/user-attachments/assets/9ea42a80-f520-4e5b-a581-75da436f9b69">

**Comment by davidkoski** [comment]
I can reproduce.  Roughly equivalent code in python does not fail:

```
import mlx.core as mx

randArr = mx.random.randint(0, 100, [1900000], dtype=mx.int64)
sortIndexes = mx.argsort(randArr, axis=-1)
sorted = randArr[sortIndexes]
mx.eval(sorted)
print(sorted.shape)
```

The code fails here:

```c++
void multi_block_sort(
...
    // Do partition
    {
      std::ostringstream kname;
      kname << "partition_mbsort_" << type_to_name(dev_vals_in) << "_"
            << type_to_name(dev_idxs_in) << "_bn" << bn << "_tn" << tn;
...
      compute_encoder.dispatchThreadgroups(grid_dims, group_dims);
```

In the python case I tried this:

```
	  std::cout << "Do partition " << kname.str() << " " << group_dims.width << " max " << kernel->maxTotalThreadsPerThreadgroup() << std::endl;
```

and see:

```
Do partition partition_mbsort_int64_uint32_bn256_tn8 929 max 1024
```

For swift I get:

```
Do partition partition_mbsort_int64_uint32_bn256_tn8 929 max 256
```

I am not sure what causes the lower limit here -- could it be other metal activity form the process (this is running in an app)?  I don't see any checks/guards to try and hit this limit.

Both python and swift using mlx v0.16.0.

@awni any thoughts?

**Comment by awni** [comment]
Oh that's odd.. maybe related to JIT compilation or a setting which is different there. Let me try some things with the JIT to see if I can get it to show up.

**Comment by awni** [comment]
We changed the source code for the [sort kernels](https://github.com/ml-explore/mlx/commits/main/mlx/backend/metal/kernels/sort.h) about a month ago. It used to hard code the max thread groups for the specific kernel that is crashing but now it doesn't.

It looks like MLX Swift is using the old [sort kernels](https://github.com/ml-explore/mlx-swift/blob/main/Source/Cmlx/mlx-generated/sort.cpp#L427-L428). The same test breaks with `mlx==0.16.0` in Python with metal validation enabled.

So in short this is fixed in the latest release in core and we should bump Swift to patch it here.

**Comment by davidkoski** [comment]
Thanks @awni !
